fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -86,470 +93,715 @@
+                    if (account.getAccountTypes() == 5) {
+                         final Long groupId = command.longValueOfParameterNamed("groupId");
+                         // GSIM specific parameters
+                         if (command.bigDecimalValueOfParameterNamedDefaultToNullIfZero("applicationId") != null) {
+                              applicationId = command.bigDecimalValueOfParameterNamedDefaultToNullIfZero("applicationId");
+                         }
+
+                         if (command.booleanObjectValueOfParameterNamed("lastApplication") != null) {
+                              isLastChildApplication = command.booleanPrimitiveValueOfParameterNamed("lastApplication");
+                         }
+
+                         Group group = this.groupRepositoryWrapper.findOneWithNotFoundDetection(groupId);
+
+                         if (command.booleanObjectValueOfParameterNamed("isParentAccount") != null)
+                         {
+                                   // empty table check
+                              if (gsimRepository.count() != 0) {
+                                   // Parent-Not an empty table
+
+                                   accountNumber = this.accountNumberGenerator.generate(account, accountNumberFormat);
+                                   account.updateAccountNo(accountNumber + "-1");
+                                   gsimAccount = gsimWritePlatformService.addGSIMAccountInfo(accountNumber, group, BigDecimal.ZERO,
+                                             Long.valueOf(1), true,
+                                             SavingsAccountStatusType.SUBMITTED_AND_PENDING_APPROVAL.getValue(),applicationId);
+                                   account.setGsim(gsimAccount);
+                                   this.savingAccountRepository.save(account);
+
+                              } else {
+                                   // Parent-empty table
+                                   accountNumber = this.accountNumberGenerator.generate(account, accountNumberFormat);
+                                   account.updateAccountNo(accountNumber + "-1");
+                                   gsimWritePlatformService.addGSIMAccountInfo(accountNumber, group, BigDecimal.ZERO,
+                                             Long.valueOf(1), true,
+                                             SavingsAccountStatusType.SUBMITTED_AND_PENDING_APPROVAL.getValue(),applicationId);
+                                   account.setGsim(gsimRepository.findOneByAccountNumber(accountNumber));
+                                   this.savingAccountRepository.save(account);
+
+                              }
+
+                         } else {
+
+                              if (gsimRepository.count() != 0) {
+                                   // Child-Not an empty table check
+                                   gsimAccount = gsimRepository.findOneByIsAcceptingChildAndApplicationId(true,applicationId);
```

Comment:
@rrpawar96 if client/front end application generates same applicationID for two applications- it would mix up the child applications.  Although the chances are very low, we should see if we can synchronize access to submitApplication() through its parent submitGSIMApplication().